### PR TITLE
support graalvm

### PR DESCRIPTION
**Pull request by joshlong**
this adds GraalVM native image configuration so that anybody that has these `.jar`s on the classpath may compile their application to a GraalVM native image with ease.

### CODE REVIEW

File: src/main/resources/META-INF/native-image/reflect-config.json [hunk 1]
```diff
@@ -0,0 +1,1127 @@
...
+    "allDeclaredMethods": true,
+    "allPublicMethods": true,
+    "allPublicConstructors": true,
+    "queryAllDeclaredMethods": true,
+    "queryAllPublicConstructors": true,
+    "allDeclaredClasses": true,
+    "allDeclaredFields": true,
+    "allPublicClasses": true,
+    "allDeclaredConstructors": true
+  },
```
Comment (linghengqian) [inline src/main/resources/META-INF/native-image/reflect-config.json:16]: - It looks like this is not generated for unit tests via the GraalVM Native Build Tools' GraalVM Agent, but handled manually?  This looks different from `oracle/graalvm-reachability-metadata`. 👀

File: src/main/resources/META-INF/native-image/resource-config.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
+{
+  "resources": {
+    "includes": [
```
Comment (andimarek) [inline src/main/resources/META-INF/native-image/resource-config.json:3]: Can you describe the purpose of these exclusions? GraphQL Java has no direct relationship with `graphiql` for example.

Comment (joshlong) [reply src/main/resources/META-INF/native-image/resource-config.json:3]: Hi! That's a good point. Feel free to not have those if you don't want. Somebody will need to provide those. Hopefully Spring Boot will do that one day soon...

File: src/main/resources/META-INF/native-image/reflect-config.json (hunk 1 above)
Comment (joshlong) [reply src/main/resources/META-INF/native-image/reflect-config.json:16]: Anybody can author the Config files; you don't need to use the java agent. I'm not sure what you mean by 'for unit tests'. These config files work in any context: I'm most interested in production code. I arrived at them by debugging a Spring for GraphQL application (and trying to make the GraphiQL console work). These Config files mean we can introspect GraphQL using GraphQL

Comment (linghengqian) [reply src/main/resources/META-INF/native-image/reflect-config.json:16]: - No problem, I just assumed commands like `./gradlew -Pagent run` and `./gradlew metadataCopy --task run --dir src/main/resources/META-INF/native-image` were used.  I understand that the metadata generated by `Conditional Agent Mode` is usually shorter, refer to https://graalvm.github.io/native-build-tools/latest/gradle-plugin.html#agent-support . 😃

**Comment by bbakerman** [comment]
I am confused here - most probably because I dont fully understand GraalVM development.  I have played with it but not in detail.

The classes listed here like `graphql.language.OperationDefinition` and the schema ones in `graphql.schema.*` are never reflected on by graphql-java code??

In fact the only reflection graphql-java does is actually onto code it does not own, eg the POJOs and records that `PropertyDataFetcher` might reach into.  Its not in graphql-java but outside it.

So I am confused why these class are listed above?  Is it because Spring Graphql reflects onto these class during its processing.

And if so should this not live in Spring?

I ask these questions from a position in inferior knowledge - I am not a Graal expert by any means

**Comment by linghengqian** [comment]
@bbakerman What do you think of the test cases provided by https://github.com/oracle/graalvm-reachability-metadata/pull/102 ?

**Comment by bbakerman** [comment]
I am going to close this issue - in other channels we have discuss how this can be approached more systematically

https://github.com/oracle/graalvm-reachability-metadata/pull/102

**Comment by bbakerman** [comment]
> @bbakerman What do you think of the test cases provided by [oracle/graalvm-reachability-metadata#102](https://github.com/oracle/graalvm-reachability-metadata/pull/102) ?

This is the way forward

**Pull request closed without merging**